src/3rd_party_bundle_add.c

```diff
@@ -94,6 +98,174 @@ static bool parse_options(int argc, char **argv)
+static bool is_binary(const char *filename)
+{
+	if (strncmp(filename, "/bin", strlen("/bin")) == 0 ||
+	    strncmp(filename, "/usr/bin", strlen("/usr/bin")) == 0 ||
+	    strncmp(filename, "/usr/local/bin", strlen("/usr/local/bin")) == 0) {
+		return true;
+	}
+
+	return false;
+}
+
+static enum swupd_code create_wrapper_script(const char *filename)
+{
+	enum swupd_code ret_code = 0;
+	FILE *fp = NULL;
+	char *bin_directory = NULL;
+	char *script = NULL;
+	char *binary = NULL;
+	char *env_path = NULL;
+	char *env_ld_path = NULL;
+	char *third_party_bin_path = NULL;
+	char *third_party_ld_path = NULL;
+	mode_t mode = 0755;
+
+	/* the path_prefix has the path to the 3rd-party repo, like:
+	 *  /<original_path_prefix>/opt/3rd-party/bundles/<repo>/
+	 *  we can use it to find the path to the bin directory:
+	 *  /<original_path_prefix>/opt/3rd-party/bin */
+	char *tmp1 = sys_dirname(globals.path_prefix);
+	char *tmp2 = sys_dirname(tmp1);
+	bin_directory = sys_path_join(tmp2, "/bin");
+	free_string(&tmp1);
+	free_string(&tmp2);
+
+	script = sys_path_join(bin_directory, sys_basename(filename));
+	binary = sys_path_join(globals.path_prefix, filename);
+
+	if (!sys_filelink_is_executable(binary)) {
+		goto close_and_exit;
+	}
+
+	if (sys_file_exists(script)) {
+		/* the binary already exists, this condition should never happen
+		 * since we checked before installing */
+		ret_code = SWUPD_UNEXPECTED_CONDITION;
+		error("There is already a binary called %s in %s, it will be skipped\n", sys_basename(filename), bin_directory);
+		goto close_and_exit;
+	}
+
+	/* if the SWUPD_3RD_PARTY_BIN_DIR does not exist, attempt to create it */
+	if (!sys_file_exists(bin_directory)) {
+		if (mkdir_p(bin_directory)) {
+			ret_code = SWUPD_COULDNT_CREATE_DIR;
+			goto close_and_exit;
+		}
+	}
+
+	if (!is_dir(bin_directory)) {
+		error("The path %s for 3rd-party content exists but is not a directory\n", bin_directory);
+		ret_code = SWUPD_COULDNT_CREATE_DIR;
+		goto close_and_exit;
+	}
+
+	fp = fopen(script, "w");
+	if (!fp) {
+		error("The file %s failed to be created\n", script);
+		ret_code = SWUPD_COULDNT_CREATE_FILE;
+		goto close_and_exit;
+	}
+
+	/* get the current values of the environment */
+	env_path = getenv("PATH");
+	env_ld_path = getenv("LD_LIBRARY_PATH");
+
+	/* get the path for the 3rd-party content */
+	third_party_bin_path = str_or_die("%sbin:%susr/bin:%susr/local/bin", globals.path_prefix, globals.path_prefix, globals.path_prefix);
+	third_party_ld_path = str_or_die("%susr/lib64:%susr/local/lib64", globals.path_prefix, globals.path_prefix);
+
+	fprintf(fp, "#!/bin/bash\n");
+	fprintf(fp, "\n");
+	fprintf(fp, "export PATH=%s%s%s\n", third_party_bin_path, env_path ? ":" : "", env_path ? env_path : "");
+	fprintf(fp, "export LD_LIBRARY_PATH=%s%s%s\n", third_party_ld_path, env_ld_path ? ":" : "", env_ld_path ? env_ld_path : "");
+	fprintf(fp, "\n");
+	fprintf(fp, "%s \"$@\"\n", binary);
+
+	/* set the file mode to 0755 */
+	if (chmod(script, mode) < 0) {
```

Comment:
Instead of setting the mode later you can use `open()` to create the file with correct permissions and have one less location to check for errors.

Reply:
`fdopen()` can help too.

Reply:
I didn't know I could use `fopen` setting the permissions at once, I will change that.

Reply:
I think you cant. You need to use `open()`. Then `fdopen()` to get the FILE*